declarative/syncprofilewatcher.cpp

```diff
@@ -0,0 +1,197 @@
+/*
+ * This file is part of buteo-syncfw package
+ *
+ * Copyright (C) 2019-2021 Damien Caliste.
+ *
+ * Contact: Damien Caliste <dcaliste@free.fr>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA
+ *
+ */
+
+#include "syncprofilewatcher.h"
+
+#include <QDebug>
+
+using namespace Buteo;
+
+SyncProfileWatcher::SyncProfileWatcher(QObject *parent)
+    : QObject(parent), mSyncProfile(nullptr), mSyncStatus(Sync::SYNC_DONE)
+{
+    connect(&mManager, &ProfileManager::signalProfileChanged,
+            this, &SyncProfileWatcher::onProfileChanged);
+    connect(&mSyncClient, &SyncClientInterface::profileChanged,
+            this, &SyncProfileWatcher::onProfileChanged);
+    connect(&mSyncClient, &SyncClientInterface::syncStatus,
+            this, &SyncProfileWatcher::onSyncStatus);
+}
+
+SyncProfileWatcher::~SyncProfileWatcher()
+{
+    delete mSyncProfile;
+}
+
+void SyncProfileWatcher::setName(const QString &aName)
+{
+    if (aName == name())
+        return;
+
+    delete mSyncProfile;
+    mSyncProfile = mManager.syncProfile(aName);
+    emit nameChanged();
+    emit displayNameChanged();
+    emit enabledChanged();
+    emit scheduleChanged();
+    emit logChanged();
```

Comment:
Up to you to consider, and perhaps here no problem, but just commonly safer if signals are emitted after side-effects. Avoids signal handlers doing strange things while there's something still ongoing.

Reply:
You're absolutely right, emission must be after effects. But here, I'm emitting signals after the internal attributes used in the corresponding getters have been set (for instance, log is depending only on mSyncProfile). So there is no ordering mistake.

Maybe, I'm misunderstanding and you suggest to move all signals after every thing have been settled, like also the status and the keys ? In case a client would fetch a key on log change for instance ?

Reply:
FWIW, at some point I started following the pattern when signals are first queued (by settings an internal bit somewhere) and then all pending signals are emitted from a consistent state, allowing signal handlers to do whatever they want with the model.

Reply:
Not saying ordering mistake, but what can sometimes happen is if you have code that set property a to something, emits signal for it, sets b to something, emits signal for it, does some changes. Then handlers for a and b might be assigning values to each other, or some other properties. If the handlers get a chance to do changes in between, the main method can then later for example overwrite the changes. Or screw up the whole state keeping. Thus safer the more there are side-effects done before the signals are emitted.

On this was just pondering would it be just slightly better if the list was last thing on the method or after setKeys() or something. The component here doesn't look too dangerous though.